HW1/P5/P5.py

```diff
@@ -0,0 +1,158 @@
+import pyspark
+sc = pyspark.SparkContext()
+
+sc.setLogLevel('WARN')
+
+
+def parse_links(x):
+    """
+    Takes arg of form "source: n1 n2 n3 n4 n5" and returns
+    (source, [n1, n2, n3, n4, n5])
+    """
+    source, neighbors = x.split(': ')
+    source = int(source)
+    neighbors = [int(i) for i in neighbors.split(' ')]
+
+    return (source, neighbors)
+
+
+links = sc.textFile('s3://Harvard-CS205/wikipedia/links-simple-sorted.txt')
+links = links.map(parse_links)
+index_to_title = sc.textFile('s3://Harvard-CS205/wikipedia/titles-sorted.txt').zipWithIndex().map(lambda x: (x[1] + 1, x[0])).cache()
+title_to_index = index_to_title.map(lambda x: (x[1], x[0])).cache()
+
+
+def larger(x, y):
+    """
+    Return larger of two lists. If tie, returns x.
+    For use in add_neighbors.
+    """
+    if len(y) > len(x):
+        return y
+    else:
+        return x
+
+
+def collapse_nodes(x, y):
+    """
+    Reducer that takes:
+        * Max of the flags
+        * Min of the distances
+        * Larger of the neighbors lists
+        * Previous of the node with the smaller distance
+    """
+    if x[3] is None:
+        previous = y[3]
+    elif y[3] is None:
+        previous = x[3]
+    elif x[1] <= y[1]:
+        previous = x[3]
+    else:
+        previous = y[3]
+
+    return (max((x[0], y[0])), min(x[1], y[1]), larger(x[2], y[2]), previous)
+
+
+def bfs(source, target):
+    """
+    Performs breadth-first search starting at source and ending once
+    target is found or all connected nodes have been touched.
+
+    Flags mean:
+    * 0 = node has not yet been reached
+    * 1 = node has been reached, but neighbors have not been marked as reached yet
+    * 2 = node has been reached and neighbors have been marked as reached
+
+    Vertices are of form (node, (flag, dist, [neighbors], previous))
+    """
+
+    def initialize_nodes(x):
+        # initialize flag to 0
+        flag = 0
+
+        # initialize distance to a high number
+        dist = 9999
```

Comment:
consider using the python construct float("inf"). dist = 9999 is more of an arbitrary workaround